main/drivers/CANDriver.c

```diff
@@ -0,0 +1,293 @@
+#include "CANDriver.h"
+
+#include <esp_log.h>
+#include <string.h>
+
+#define CAN_STACK_SIZE 2048 // Minimum
+
+static const char *TAG = "CANDriver";
+
+static CANDriver_t driver;
+
+/**
+ * The ISR handler for the INT1 pin of device triggered on rising edge
+ * @param arg The can bus number
+ */
+void IRAM_ATTR CANDriverISRHandler(void *arg) {
+    uint32_t bus = (uint32_t)arg;
+
+    // set the corrsponding event to chip with data (this basically defers
+    // the SPI reading to a later time that is not in an interrupt).
+    xEventGroupSetBitsFromISR(driver.messageEvents, 1 << bus, pdTRUE);
+}
+
+/**
+ * The FreeRTOS task for reading
+ * @param arg UNUSED (part to FreeRTOS spec)
+ */
+static void CANDriverTask(void *arg) {
+    EventBits_t activeEvents = 0;
+
+    // wait until the mask has atleast on bit set
+    while (!driver.deviceEventMask)
+        vTaskDelay(1);
+
+    for (;;) {
+        // determine which bus has just got data
+        if (activeEvents) {
+            // get any additional events while there are still messages in the FIFO
+            activeEvents |= xEventGroupClearBits(driver.messageEvents, driver.deviceEventMask);
+        } else {
+            // get any evens, wait until an even happens or 10ms without event
```

Comment:
The interupt that sets the flag only occures on the falling edge. In addition this task is responsible for recieving messages from all of the CAN chips (since all chips are on the same bus and to prevent task contention for the bus). The `deviceEventMask` simply says which CAN bus device to listen for events from.